### PR TITLE
WIP integration with FusionAuth

### PR DESCRIPTION
**Pull request by macklinu**
The code we explored today with FusionAuth and Next Auth. Let me know if the instructions are unclear in the README for getting some of this stuff set up.

### CODE REVIEW

**Comment by JoshLinneburg** [comment]
Seems like I managed to find a solution to our issues here. I'm not sure if this is 100% correct, but the app is working as intended now after I made some changes. 
I'll try to break down what I found doing some reading last night:

1. According to [this comment on a Next-Auth GitHub issue](https://github.com/nextauthjs/next-auth/discussions/1039#discussioncomment-260223), the `JWT_SECRET` env. variable "is used to generate a key suitable for signing but isn't really the key itself - it's more a value used to seed a derived key."
2. The output of the `jose newkey -s 512 -t oct -a HS512` command we were running *does* contain the secret key in the "k" key: value pair - but the default behavior of Next-Auth does not seem to use this key directly. 
3. Instead, it appears like Next-Auth base64 encodes this key before signing the JWT. See below:

Here are two screenshots of the same JWT being decoded with the same secret key (plucked from the `jose newkey` command output) on jwt.io. 
[In the first](https://raw.githubusercontent.com/JoshLinneburg/react-tutorial/fusionauth/screenshots/invalid-jwt.png), I don't check the "secret base64 encoded" box in the "Verify Signature" section. jwt.io returns an "invalid signature" error. 
[In the second](https://raw.githubusercontent.com/JoshLinneburg/react-tutorial/fusionauth/screenshots/verified-jwt.png), I do check this box and pass in the exact same secret. This time, jwt.io is able to successfully decode the JWT. 

I tried to base64 encode the "real" secret key in Python and decode w/ that, specifying my own `getSigningKey` function like the original GitHub comment I referenced does, etc. But I wasn't able to get anything to work (maybe I need to do some more reading/tinkering) and it all felt pretty convoluted when all we want to do is pass a JWT between Next-Auth and Flask, and decode with an arbitrary secret key.  

So, I just followed [this guy's](https://github.com/nextauthjs/next-auth/issues/716) lead and specified my own `encode` and `decode` functions for Next-Auth. No base64 encoding, just use the `JWT_SECRET` env. variable as we originally thought it was being used (i.e., as the secret key itself, not some "seed" to set a secret key), sign the JWT, and pass it along. On the Flask side, I set the `JWT_SECRET_KEY` env. variable to the same value and voilà, everything works. 

I can now login to the application we were working on and when I do, the endpoint allows me to retrieve the user data on the homepage as intended. No more 422's from Flask for being unable to verify the signature of the JWT.

File: README.md [hunk 1]
```diff
@@ -11,23 +12,61 @@ pip install -r requirements.txt
...
+# The application's client secret
+FUSIONAUTH_CLIENT_SECRET='FILL_IN'
+# The FusionAuth API key created in the previous step
+FUSIONAUTH_API_KEY='FILL_IN'
+FLASK_API_BASE_URL='http://localhost:5000'
+# A JWT secret string - see the Next Auth docs: https://next-auth.js.org/configuration/options#jwt
+JWT_SECRET='FILL_IN'
+```
+
+It is still unclear, but you may also need to update `api/.flaskenv` with a matching `SECRET_KEY` to what is used by Next Auth. We are still trying to figure that piece out though.
```
Comment (macklinu) [inline README.md]: Your findings indicate that the web app's `JWT_SECRET` and Flask app's `SECRET_KEY` are identical. I confirmed that behavior locally (nice job!). We can update this line to say something like:

```suggestion
The `JWT_SECRET` in `.env` and `SECRET_KEY` in `api/.flaskenv` should be identical and set to the output of `jose newkey -s 512 -t oct -a HS512`. 
```

We could even make the environment variable names the same if we wanted (`SECRET_KEY` or `JWT_SECRET`) - sort of minor detail, but something to consider.

File: api/flask_app/users/routes.py [hunk 2]
```diff
@@ -4,16 +4,27 @@
...
 
 @users_bp.route("", methods=["GET"])
+@jwt_required()
 def read_users():
+    headers = request.headers
+    access_token = headers.get("Authorization").split(" ")[-1]
+
+    try:
+        decoded_token = decode_token(encoded_token=access_token)
+        pprint(decoded_token)
```
Comment (macklinu) [inline api/flask_app/users/routes.py:24]: ℹ️ `get_jwt_identity()` will return the same value of `decoded_token["sub"]` if we ever just need the user ID to perform a lookup in our own database / FusionAuth's API for authorization purposes.

**Review by macklinu — COMMENTED** [review]
Nice job on figuring this one out! This seems like a good approach. I think next steps would be using the decoded token / token's sub (user ID) to then ensure the user making the API request has the appropriate permissions to perform operations. Example: you may want all users to be able to GET /api/v1/user, but only admins can POST or PUT to /api/v1/user. Overall, this seems really close to being at a point where we can build a full prototype with secure APIs.

**Review by JoshLinneburg — APPROVED** [review]

**Pull request merged**